docs/user-guide/timeouts/timeouts-main.md

```diff
@@ -280,28 +302,80 @@ certoraRun ... --prover_args '-dontStopAtFirstSplitTimeout true -depth 15 -mediu
+solved. Nevertheless, reducing the number of nonlinear operations has often
+proven a successful measure in timeout prevention even if some remained.
+```
+
+The main techniques in reducing these numbers are modularization and
+underapproximation. 
 
-% For instance, we can prioritize the usage of the [Yices SMT
-% solver](https://yices.csl.sri.com/) by decreasing the size of the solver
-% portfolio. With the {ref}`-solver` option set as follows, the Certora Prover
-% will run only CVC5 and Yices. Furthermore, we can make the Certora Prover use
-% the ordering given in the {ref}`-solver` option for prioritizing solvers using
-% the `-smt_overrideSolvers` option.
+Modularization, typically by introducing method summaries, can help reduce the
+size of the rule, thus reducing the nonlinear operations. The per-call
+statistics in the Live Statistics panel (picture below) can help with
+identifying nonlinearity hot spots. Summarizing these hot spots in particular
+can help reduce the number of nonlinear operations, especially when a method is 
+called multiple times.
 
-% TODO make this subsection a bit more concrete.. not sure how, yet
+```{figure} nonlinear-ops-call.png
+:name: nonlinear-ops-call
+Entry in Live Statistics indicating how many nonlinear operations are made in a given 
+call, including its sub-calls
+```
 
-% ```sh
-% certoraRun ... --prover_args '-solvers [yices, cvc5] -smt_overrideSolvers true'
-% ```
+In some rules it is feasible to only consider an underapproximation of the
+actual behavior by fixing some value that is used very often in nonlinear
+computations to a concrete value. A typical example would be the decimal digits
+in fixed decimal arithmetic -- having this unconstrained can increase
+nonlinearity in the rule massively, although only a small range of values is
+actually feasible. Of course, great care has to be taken in choosing these
+underapproximations, since they lead to missed bugs otherwise.
+
+A weaker form of underapproximation would be to introduce an extra requirement
+on the range of some variable that contributes to nonlinearity. For example for
+the number of decimals in a fixed decimal computation only values between 0 and
+256 make sense, and in practice values from an even smaller range are likely to
+be used. This measure will not change the values in the Live Statistics panel, 
+but it has prevented timeouts in some cases nonetheless.
 
 (high-memory-complexity)=
 ### Dealing with high memory (or storage) complexity
 
-In this subsection we consider common culprits for high memory complexity.
+The memory complexity of each rule or parametric rule is displayed in the Live
+Statistics panel in the Certora Prover reports. Memory complexity is measured by
+two values *graph size*, and *longest path*. Graph size indicates how often
+memory is updated in the rule overall.  Longest path indicates how many updates
+any given internal memory variable (e.g. the storage of a given contract, or a
```

Comment:
I'm totally confused now. so what's the graph and the path. I was under the impression that it is dependencies between updates, i.e, def-use chains. Is it just number of operations?

Reply:
ugh.. maybe we should drop the mem complexity heuristic altogether ... 
or have just one number 🤔 
or maybe explain less?

I'll sleep on it ..

Reply:
actually it's useful. I understand it like this, feel free to rephrase or use:

> The Certora Prover performs decompilation of bytecode converting a stack-based machine to a register-based machine, that includes key-to-value mappings for EVM objects such as memory and storage. In some cases, it's able to simplify the mappings to use simpler register variables. 
> 
> Under this model, the "Graph size" is a measure of how many times we store into a key-value mapping such as memory, storage, or a ghost function.
> The "Path" is a measure of the longest chain of writes performed on one of the mappings.
> In both cases, a smaller number indicates a less difficult problem for the Prover to solve.

Reply:
used Shelly's text (hope I didn't butcher it too much ... or only in a good way :-) ) -- keeping the stat as is for now